### PR TITLE
Connect frontend to backend endpoints

### PR DESCRIPTION
**Pull request by ErlingHauan**
## FormViewer
* When visiting "localhost/view/some-unique-id", the frontend will now display the corresponding form with id "some-unique-id", fetched from the database.
* When making a submission from the form viewer, the submission will be stored in the database.

## Dashboard
* The dashboard statistics will count the number of forms and submissions in the database.
* The actual forms from the database will now be displayed.
* Made DashboardOverview and DashboardAccordion separate components from Dashboard.

## SubmissionViewer
* The actual submissions from the database will now be displayed.

### CODE REVIEW

File: frontend/main-app/src/components/FormViewer/FormViewer.tsx [hunk 1]
```diff
@@ -22,30 +30,46 @@ export const FormViewer = (): React.JSX.Element => {
...
+            </Button>
+          </div>
+        </form>
+        {formAlert == "success" && <Alert severity="success">{t("form_viewer.success")}</Alert>}
+        {formAlert == "validationError" && (
+          <Alert severity="danger">{t("form_viewer.validationError")}</Alert>
+        )}
+        {formAlert == "serverError" && (
+          <Alert severity="danger">{t("form_viewer.serverError")}</Alert>
+        )}
```
Comment (lassopicasso) [inline frontend/main-app/src/components/FormViewer/FormViewer.tsx:61]: Can you separate this logic to a reusable function? And invoke it with args based on the type of `formAlert`.

File: frontend/main-app/src/components/FormViewer/FormViewer.tsx [hunk 2]
```diff
@@ -22,30 +30,46 @@ export const FormViewer = (): React.JSX.Element => {
...
-            Submit form
-            <TasklistSendFillIcon />
-          </Button>
-        </div>
-      </form>
-      {formAlert == "success" && <Alert severity="success">{t("form_viewer.success")}</Alert>}
-      {formAlert == "error" && <Alert severity="danger">{t("form_viewer.error")}</Alert>}
-    </main>
-  );
+  if (formSchema) {
```
Comment (lassopicasso) [inline frontend/main-app/src/components/FormViewer/FormViewer.tsx:40]: What do you think about extracting this into two blocks instead?

```
    return (
      <main className={classes.card}>
        {formSchema ?
           <RenderForm ... />
         : <RenderInvalidForm .../>
      </main>
      )
      
```

File: frontend/main-app/src/components/FormViewer/FormViewer.tsx [hunk 3]
```diff
@@ -22,30 +30,46 @@ export const FormViewer = (): React.JSX.Element => {
...
+      </main>
+    );
+  }
+
+  if (!formSchema) {
+    return (
+      <main className={classes.card}>
+        <Heading spacing>Form not found</Heading>
+        {formId && <Paragraph>Could not find form with ID {formId}.</Paragraph>}
+        {!formId && <Paragraph>No form Id was provided in the URL.</Paragraph>}
```
Comment (lassopicasso) [inline frontend/main-app/src/components/FormViewer/FormViewer.tsx:71]: Maybe a similar approach as the suggested change for `formAlert`?

File: frontend/main-app/src/components/FormViewer/httpUtils.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    responses.push(response);
+    i++;
+  }
+
+  return {
+    formId: formSchema.id,
+    submitted: new Date(),
+    responses: responses,
+  };
+};
```
Comment (lassopicasso) [inline frontend/main-app/src/components/FormViewer/httpUtils.tsx:38]: Your solution is good, but I think utilizing the mapping's potential here it makes it even more readable.
Haven't tested that this works, so maybe I missing something here.
```suggestion
const formatSubmission = (formSchema, formData) => {
  const responses = formSchema.components.map((component, index) => ({
    name: component.name,
    order: component.order,
    label: component.label,
    response: formData[index],
  }));

  return {
    formId: formSchema.id,
    submitted: new Date(),
    responses,
  };
};
```

File: frontend/main-app/src/components/FormViewer/httpUtils.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  const targetUrl = `${apiUrl}/api/submissions/`;
+
+  try {
+    const response = await axios.post(targetUrl, formattedSubmission);
+    setFormAlert("success");
+    console.log(response.status);
+  } catch (error) {
+    setFormAlert("serverError");
+    console.log(error);
+  }
```
Comment (lassopicasso) [inline frontend/main-app/src/components/FormViewer/httpUtils.tsx:52]: I guess you don't need the console.log when success 😄 
```suggestion
  try {
    const response = await axios.post(targetUrl, formattedSubmission);
    setFormAlert("success");
  } catch (error) {
    setFormAlert("serverError");
    console.log(error);
  }
```

**Review by lassopicasso — CHANGES_REQUESTED** [review]
Solid work! I have some thoughts that you can read below.

File: frontend/main-app/src/components/FormComponents/FormComponents.tsx [hunk 6]
```diff
@@ -2,28 +2,32 @@ import classes from "./FormComponents.module.css";
...
   errors,
 }): React.JSX.Element => {
   return (
     <>
       {components.map((component: FormComponent) => {
-        const { componentType, name, label, required, choices } = component;
+        const { type, name, label, required, radioChoices } = component;
 
-        return componentType === "textfield" ? (
+        return type === "textfield" ? (
```
[the remark this answers is not in the document]
Comment (ErlingHauan) [reply frontend/main-app/src/components/FormComponents/FormComponents.tsx]: Thanks for the suggestion. I used the last one, removing the logic outside the return statement.

File: frontend/main-app/src/components/FormViewer/FormViewer.tsx (hunk 1 above)
Comment (ErlingHauan) [reply frontend/main-app/src/components/FormViewer/FormViewer.tsx]: Yes 😊
````
export const alertToRender = (alertType: string, t: TFunction): React.JSX.Element => {
  if (alertType == "success") {
    return <Alert severity="success">{t("form_viewer.success")}</Alert>;
  } else if (alertType == "validationError") {
    return <Alert severity="danger">{t("form_viewer.validationError")}</Alert>;
  } else if (alertType == "serverError") {
    return <Alert severity="danger">{t("form_viewer.serverError")}</Alert>;
  }
};
````

File: frontend/main-app/src/components/FormViewer/FormViewer.tsx (hunk 2 above)
Comment (ErlingHauan) [reply frontend/main-app/src/components/FormViewer/FormViewer.tsx]: Yes, that might be easier to read. Fixed ✅

File: frontend/main-app/src/components/FormViewer/FormViewer.tsx (hunk 3 above)
Comment (ErlingHauan) [reply frontend/main-app/src/components/FormViewer/FormViewer.tsx]: ✅ 
```
export const getFormIdError = (formId: string, t: TFunction) => {
  if (formId) {
    return (
      <Paragraph>
        {t("form_viewer.formId.notFound")}
        {formId}.
      </Paragraph>
    );
  }
  return <Paragraph>{t("form_viewer.formId.notProvided")}</Paragraph>;
};
```

File: frontend/main-app/src/components/FormViewer/httpUtils.tsx (hunk 4 above)
Comment (ErlingHauan) [reply frontend/main-app/src/components/FormViewer/httpUtils.tsx]: That's really smooth, much more readable! 💯 
This is how it turned out after some testing:

````
const formatSubmission = (formSchema, formData) => {
  const responses = formSchema.components.map((component) => ({
    name: component.name,
    order: component.order,
    label: component.label,
    response: formData[component.name] || "No response",
  }));

  return {
    formId: formSchema.id,
    submitted: new Date(),
    responses,
  };
};
````

File: frontend/main-app/src/components/FormComponents/FormComponents.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (lassopicasso) [reply frontend/main-app/src/components/FormComponents/FormComponents.tsx]: Super :)

**Pull request merged**